rust/agama-lib/src/file_source.rs

```diff
@@ -0,0 +1,186 @@
+// Software Foundation; either version 2 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, contact SUSE LLC.
+//
+// To contact SUSE LLC about this file by physical or electronic mail, you may
+// find current contact information at www.suse.com.
+
+use crate::utils::{Transfer, TransferError};
+use fluent_uri::{error::ResolveError, Uri, UriRef};
+use serde::{Deserialize, Serialize};
+use std::{fs::OpenOptions, io::Write, os::unix::fs::OpenOptionsExt, path::Path};
+
+#[derive(Debug, thiserror::Error)]
+pub enum FileSourceError {
+    #[error("Could not resolve the URL: {0}")]
+    ResolveUrlError(String, #[source] ResolveError),
+    #[error("Transfer error: {0}")]
+    TransferFailed(#[from] TransferError),
+    #[error("I/O error: {0}")]
+    IO(#[from] std::io::Error),
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, utoipa::ToSchema)]
+#[serde(untagged)]
+pub enum FileSource {
+    /// File content.
+    Text { content: String },
+    /// URI or relative reference to get the script from.
+    Remote {
+        #[schema(value_type = String, examples("http://example.com/script.sh", "/file.txt"))]
+        url: UriRef<String>,
+    },
+}
+
+impl FileSource {
+    /// Returns a new source using an absolute URL if it was using a relative one.
+    ///
+    /// If it was not using a relative URL, it just returns a clone.
+    ///
+    /// * `base`: base URL.
+    pub fn resolve_url(&self, base: &Uri<String>) -> Result<FileSource, FileSourceError> {
+        let resolved = match self {
+            Self::Text { content } => Self::Text {
+                content: content.clone(),
+            },
+            Self::Remote { url } => {
+                let resolved_url = if url.has_scheme() {
+                    url.clone()
+                } else {
+                    let resolved = url
+                        .resolve_against(base)
+                        .map_err(|e| FileSourceError::ResolveUrlError(url.to_string(), e))?;
+                    UriRef::parse(resolved.to_string()).unwrap()
+                };
+                Self::Remote { url: resolved_url }
+            }
+        };
+        Ok(resolved)
+    }
+
+    /// Writes the file to the given writer.
+    ///
+    /// * `file`: where to write the data.
+    pub fn write<P: AsRef<Path>>(&self, path: P, mode: u32) -> Result<(), FileSourceError> {
+        let mut file = OpenOptions::new()
+            .mode(mode)
+            .write(true)
+            .create(true)
+            .open(path)?;
+
+        match &self {
+            FileSource::Text { content } => write!(file, "{}", &content)?,
```

Comment:
NP: (and untested)
formatted write seems inappropriate here

```suggestion
            FileSource::Text { content } => file.write_all(content.as_bytes())?,
```